### ISSUE TITLE
[BUG] Gemini Pro 002 doubling last sentences of output.

### ISSUE THREAD

**Issue by Ph0rk0z** [SillyTavern/SillyTavern#2960]
### Environment

🪟 Linux

### System

Firefox or Chrome on Arch

### Version

staging/head

### Desktop Information

Gemini API

### Describe the problem

A couple of days ago I started seeing the last sentence in Gemin Pro 002 double. Perhaps they changed something in how the reply comes back? I can switch to Pro 001 and refresh and it will be gone in some replies but then come back. I can edit the doubled lines out, but it's annoying. Tried to disable streaming but did not fix. Happens to both prompts that reply long and ones that are short.

Not sure how else to investigate it but I have a screenshot. It's not happening on openrouter or locally so must be gemini specific.

### Additional info

![gemini-doubling](https://github.com/user-attachments/assets/876b9597-7a91-4451-8296-b69b7d817f19)

### Please tick the boxes

- [X] I have explained the issue clearly, and I included all relevant info
- [X] I have checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] I have checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**Comment by Noltron** [comment]
I got the exact same issue, It happened yesterday

**Comment by Ph0rk0z** [comment]
There are 3 of us. It's platform agnostic: https://github.com/SillyTavern/SillyTavern/issues/2961

**Comment by Cohee1207** [comment]
I wasn't able to replicate this yet. Are there any specific circumstances when this occurs?

**Comment by Ph0rk0z** [comment]
Nothing special. Just using it as normal. All gemini pro 1.5 seem to do it but the old 1.0 pro and flash do not.

![settings](https://github.com/user-attachments/assets/d58efdc9-29c9-4014-833b-463c7187f2be)

![st-repeat](https://github.com/user-attachments/assets/df3330a3-9bce-4bb0-b1d6-f7c5ac722d66)

**Comment by KingXMAX** [comment]
Sadly, the repetition problem has expanded to other models. Command R, R Plus, Mistral Large and even AI21 Jamba Large are now plagued with the repetition problem no matter what the prompt says or how the parameters are configured. Something has to be done here. Seems like you have quite a bit of work to get done, Cohee.

**Comment by Cohee1207** [comment]
Despite all my attempts, I wasn't able to replicate just yet. Cohere, Mistral, Google API types, streaming (including smooth) and non-streaming.
If you're using any unofficial backends, please tell right away, maybe it's not an ST problem.

**Comment by KingXMAX** [comment]
I use Gemini AI studio API keys, official API keys for Cohere, same for Mistral and AI21. However, they all give repeated responses when i swipe for another response, and Gemini gives cut-off ones in addition to repetitions. Do you have any settings that could prevent such issues? I mean ones other than the prompt and parameters.

**Comment by Ph0rk0z** [comment]
Could it be some NPM issue? Like with a package? I recently upgraded arch again before this started. I will check cohere but I don't have mistral. I have had no problems using openrouter.

**Comment by KingXMAX** [comment]
@Ph0rk0z Openrouter is fine and gives no issues from my tests, and as for NPM, i upgraded it to 10.9.0 yesterday but the problem is still here.

**Comment by Ph0rk0z** [comment]
Cohere working fine here so far. No repetition, the swipes are different.

**Comment by KingXMAX** [comment]
@Ph0rk0z Did you try both versions of CR and CR+, the August versions?

**Comment by Ph0rk0z** [comment]
I just did the original CR+, I can try the august one too.

**Comment by KingXMAX** [comment]
Well, i have done some tests a couple of minutes ago on CR and CR+ (both the original and august versions) and it seems that repetition is much less when i use a different API key from a different account, but the repetitions are now random, not regular.

**Comment by Cohee1207** [comment]
> Could it be some NPM issue? Like with a package? 

Doubt, but can you try on the release branch? Does the issue replicate there, or only on staging?

**Comment by Ph0rk0z** [comment]
I tested all CR+ and it doesn't repeat itself, especially like this. It's going to be hard to test release because all my settings and stuff has been from staging. If I switch the branch it might cause some problems with settings? 

Possible to see the raw message back from gemini in console or browser console? Normally I only see what I sent to it until I send another message. That way we'll know if it happens pre or post parsing.

**Comment by KingXMAX** [comment]
@Ph0rk0z Since the repetition issue can be less on Cohere when using a different API key, we can now focus on the hardest part, Gemini and its Flash and Pro models, thats where the pain is.

**Comment by Cohee1207** [comment]
> Possible to see the raw message back from gemini in console or browser console? Normally I only see what I sent to it until I send another message. That way we'll know if it happens pre or post parsing.

Try disabling streaming and look into the server console.

**Comment by Ph0rk0z** [comment]
Hmm.. after doing that again it stopped until I turned streaming back on. We're getting closer. I could swear it was the first thing I tried.

**Comment by KingXMAX** [comment]
I also tried Gemini without streaming, and it gives "Candidate text is empty" red error box, not sure why, but it doesn't give this error when the streaming is on.

**Comment by Cohee1207** [comment]
"Candidate text is empty" mainly happens when your prompt is getting block out by google (it usually gives a reason for that in the response data). I think the problem with google streaming data might be because it tries to apply the output redaction live, but ST doesn't support that, and instead duplicates the line.

**Comment by Ph0rk0z** [comment]
In my case I don't have problems with it redacting. It happens even on coding prompts. When it censors, the whole reply is cut.

**Comment by Cohee1207** [comment]
Still nothing on my end, sadly. Was not able to make it duplicate even once. I would want to see the streaming chunks for any anomalies, but you'd need to patch openai.js to add the console.log on your end.

```diff
--- a/public/scripts/openai.js
+++ b/public/scripts/openai.js
@@ -1933,6 +1933,7 @@ async function sendOpenAIRequest(type, messages, signal) {
                 const { done, value } = await reader.read();
                 if (done) return;
                 const rawData = value.data;
+                console.log(rawData);
                 if (rawData === '[DONE]') return;
                 tryParseStreamingError(response, rawData);
                 const parsed = JSON.parse(rawData);
```

@Ph0rk0z can you do it on your end and then share the browser console logs?

**Comment by Ph0rk0z** [comment]
The last chunk comes twice?

![gemini-log](https://github.com/user-attachments/assets/7604cab5-d0c0-44af-b7ea-9af454183118)

**Comment by Cohee1207** [comment]
Interesting. And I see that both chunks equally increment the completion token count. And there is nothing in the prompt data that could let us distinguish between them, and they got the same index. Overall, I think that it's a bug on Google side. We just show what arrives in the replies.

**Comment by Ph0rk0z** [comment]
Guess I'm stuck leaving streaming off for now.

**Comment by KingXMAX** [comment]
I have some interesting findings after some tests i did today. Repetition issue in other models (except for AI21 Jamba Large) seems to fade away when using a different API key AND clearing the entire chat history and starting the story from scratch. As for Gemini's current issues, someone should report them to the official google ai thread. This will shed their light on the issue so they can resolve it quickly. I hope that helps!

**Comment by KingXMAX** [comment]
Urgent Update: I have spent 3 hours testing ST and another 3 hours with RisuAI, tested the majority of Chat Completion models, different prompts, different API keys and different parameters configuration on both ST and RisuAI, and i was amazed by how Risu performs, especially with all Gemini models. With Risu, there are no repetitions, no cut-offs whatsoever even if the prompt is explicit, while we are here, stuck with repetitions in ST. I told you from the beginning that there is something wrong with ST. You want an evidence? Fine! Go ahead and test RisuAI yourself with a well-written prompt and optimal parameters and see what i'm talking about.

**Comment by Ph0rk0z** [comment]
I don't get much repetition, repetition in gemini. Only 1 or 2 characters start repeating rewordings of the last sentence deeper in the chat. The model doesn't have any visible rep penalty. So that's 100% on google.

In this case, I turned off streaming and it works for now. It gens fast enough. Maybe I'll give risu a go again sometime, but last I used it, it was *too* roleplay centric. It was like a visual novel but I like a mix. More of a messenger than a videogame. For fuller immersion I want something like amica, but my client is too slow to use the live2d without excessive CPU utilization. There's also another paid UI that lets the model control the 3d "person". I bet that's neat.

Plus gemini is only one of the API I have, along with local models. It's just the biggest model and for me isn't filtered. Would be interesting to edit risu to dump it's stream chunks and see if they don't have the problem and then look at the code they use to see if they do something different or if they compare at the end and remove doubled chunks.

**Comment by KingXMAX** [comment]
Yes, it's streaming that might be part of the problem. RisuAI doesn't have streaming function in Gemini, and thus its working flawlessly without issues even with NSFW prompts, but in ST, the response gets cut off if the prompt is explicit and the streaming is enabled, and when trying the same prompt with the streaming disabled, it gives "Candidate text is empty" error. I think the code of RisuAI must be analyzed and then fixing SillyTavern accordingly, that might be the only way out.
